app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRemoteRepository.kt

```diff
@@ -0,0 +1,77 @@
+package org.p2p.wallet.moonpay.repository.sell
+
+import org.p2p.core.token.Token
+import org.p2p.core.utils.isMoreThan
+import org.p2p.core.utils.scaleShort
+import org.p2p.wallet.common.di.AppScope
+import org.p2p.wallet.common.feature_toggles.toggles.remote.SellEnabledFeatureToggle
+import org.p2p.wallet.home.repository.HomeLocalRepository
+import org.p2p.wallet.infrastructure.dispatchers.CoroutineDispatchers
+import org.p2p.wallet.moonpay.api.MoonpayApi
+import org.p2p.wallet.moonpay.api.MoonpayIpAddressResponse
+import timber.log.Timber
+import java.math.BigDecimal
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.mapNotNull
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+private const val TAG = "MoonpaySellRemoteRepository"
+
+class MoonpaySellRemoteRepository(
+    private val moonpayApi: MoonpayApi,
+    private val sellFeatureToggle: SellEnabledFeatureToggle,
+    private val moonpayApiKey: String,
+    private val homeLocalRepository: HomeLocalRepository,
+    private val dispatchers: CoroutineDispatchers,
+    appScope: AppScope
+) : MoonpaySellRepository {
+
+    private class MoonpayRepositoryInternalError(override val cause: Throwable) : Throwable(cause.message)
+
+    // todo: maybe extract caching flags to a separate repository to reuse
+    private var cachedMoonpayIpFlags: MoonpayIpAddressResponse? = null
+    private var isUserBalancePositive: Boolean = false
+        set(value) {
+            field = value
+            Timber.tag(TAG).i("isUserBalancePositive updated to: $field")
+        }
+
+    init {
+        appScope.launch {
```

Comment:
didn't find a correct way to listen to user balance without making isSellAllowedForUser suspend